### PR TITLE
Move from mdbook to Material for Mkdocs

### PR DESCRIPTION
**Pull request by c-peters**
### Background

Initially this started out more as a proof of concept, but I ended up adding more and more that at some point I thought why not finish. I know this is a massive PR and does not really follow the guideline of 'change one thing at a time'. But hopefully in the end it will all be worth it.

### Goal

I would like to move away from mdbook for our user guide mainly as it is visually not very appealing and its functionality is limited and rust oriented. 

[Material for Mkdocs](https://squidfunk.github.io/mkdocs-material/) is a mature project with plenty of [high level backers](https://squidfunk.github.io/mkdocs-material/insiders/). It is open source and freely available. It is build on top of mkdocs with a rich ecosystem of plugins (e.g. macros, execution , code snippets, ...) and extended markdown support. It looks amazing and so far in my opinion very easy to work with.

### About this PR

I have tried to move the content from the current User Guide as is, but some things were outdated/misplaced or incomplete so I made some adjustments. The main thing I added was a concepts chapter where I explain a bit about Polars functionality from a more design perspective. You can view the current state at [https://c-peters.github.io/polars-book/](https://c-peters.github.io/polars-book/)

Some design choices / remarks:
- Separated the getting started from the User guide for clarity
- Example code is separated from markdown and inserted using [macro's](https://mkdocs-macros-plugin.readthedocs.io/en/latest/pages/) and [Snippets](https://facelessuser.github.io/pymdown-extensions/extensions/snippets/)
- Example code is executed and output is inserted in markdown (no static results)
- Functions in the code examples are linked to their corresponding API documentation for better navigation
- Python / Node code is run as a Github Action to ensure it is working with the latest version of Polars (Rust is still todo)
- Markdown links are checked as a Github Action

There are somethings still under construction and things to do (see below), but I am chasing a moving target at the moment. The state is on par functionality wise with the current user guide in my opinion and if possible I would like to merge it before adding more stuff and making the PR even more complicated.

### Todo in the future 

- [] Incorporate when to activate which feature flags in the code snippets 
- [] Write up pages still under construction (Casting, Streaming, Strings, ...)
- [] Write examples in rust / NodeJs
- [] Run Rust Example code snippets in Github Action
- [] Add a section on the Categorical Data Type
- [] Add a section / chapter on plotting
- [] Better explanation of the code_block macro in the code / contribution guide. 

There is probably a lot more that I am forgetting.

### CODE REVIEW

**Comment by ritchie46** [comment]
Cool stuff!

I think we just have to merge as this is so much. I will take a look visually to the end result and if you think it's ready, I will merge.

P.S. Have you got an idea how much PR's this is behind main? I think we should merge and re-apply those PR's that are behind.

File: docs/snippets/rust_lazy_clone_warning.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+!!! rust " Rust Users Only"
+
+    Due to historical reasons the eager API in Rust is outdated. In the future we would like to redesign it as a small wrapper around the lazy API (as is the design in Python / NodeJS). In the  examples we will use the lazy API instead with `.lazy()` and `.collect()`. For now you can ignore these two functions. If you want to know more about the lazy and eager API go [here](../user-guide/concepts/lazy-vs-eager.md).
```
Comment (ritchie46) [inline docs/snippets/rust_lazy_clone_warning.md:3]: Good to note this!

File: docs/snippets/rust_lazy_clone_warning.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+!!! rust " Rust Users Only"
+
+    Due to historical reasons the eager API in Rust is outdated. In the future we would like to redesign it as a small wrapper around the lazy API (as is the design in Python / NodeJS). In the  examples we will use the lazy API instead with `.lazy()` and `.collect()`. For now you can ignore these two functions. If you want to know more about the lazy and eager API go [here](../user-guide/concepts/lazy-vs-eager.md). 
+
+    Because of the ownership ruling in Rust we can not reuse the same `DataFrame` multiple times in the examples. For simplicity reasons we call `clone()` to overcome this issue. Note that this does not duplicate the data but just creates a pointer in memory (`Arc`)
```
Comment (ritchie46) [inline docs/snippets/rust_lazy_clone_warning.md:5]: > but just creates a pointer in memory (`Arc`)

I would rephrase this as it `"just increments a pointer"`

File: docs/src/data/appleStock.csv [hunk 3]
```diff
@@ -0,0 +1,9 @@
+Date,Close
+1981-02-23,24.62
+1981-05-06,27.38
+1981-05-18,28.0
+1981-09-25,14.25
+2012-12-04,575.85
+2013-07-05,417.42
+2013-11-07,512.49
+2014-02-25,522.06
```
Comment (ritchie46) [inline docs/src/data/appleStock.csv:9]: leading whitespace?

File: docs/src/node/getting-started/expressions.js [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+// --8<-- [end:combine]
+
+// --8<-- [start:combine2]
+df_y = df.withColumns([
+    (pl.col("a").mul(pl.col("b"))).alias("a * b")
+]).select([
+    pl.all().exclude('d')
+])
+console.log(df_y)
+// --8<-- [end:combine2]
```
Comment (ritchie46) [inline docs/src/node/getting-started/expressions.js:98]: Do these trailing new line chars remain after a black run?

File: docs/user-guide/concepts/data-structures.md [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+{{code_block('getting-started/series-dataframes','series',['Series'])}}
+
+```python exec="on" result="text" session="getting-started/series"
+--8<-- "python/getting-started/series-dataframes.py:series"
+```
+
+## DataFrame
+
+A `DataFrame` is a 2-dimensional data structure that is backed by a `Series`, and it could be seen as an abstraction of on collection (e.g. list) of `Series`. Operations that can be executed on `DataFrame` are very similar to what is done in a `SQL` like query. You can `GROUP BY`, `JOIN`, `PIVOT`, but also define custom functions.
```
Comment (ritchie46) [inline docs/user-guide/concepts/data-structures.md:18]: abstraction of ~on~ a collection

**Comment by c-peters** [comment]
> Cool stuff!
> 
> I think we just have to merge as this is so much. I will take a look visually to the end result and if you think it's ready, I will merge.
> 
> P.S. Have you got an idea how much PR's this is behind main? I think we should merge and re-apply those PR's that are behind.

Do you mean open PR's or closed ones? 
For the closed ones, I have incorporated and re-applied them already.

**Comment by c-peters** [comment]
Do you have the actual appleStock.csv that is being used currently for me?
Right now I created my own with a few rows.

**Comment by ritchie46** [comment]
@c-peters the applestock data seems to be fetched: https://github.com/pola-rs/polars-book/blob/278a599e5be1b9a794068877c09cc2367103fec6/Makefile#L15

**Comment by alexander-beedie** [comment]
Typo ([intro page](https://c-peters.github.io/polars-book/)):
`provide meaningfull examples` → `provide meaningful examples`
 
For a future/follow-up PR... :)

- [ ] Dark mode CSS/theme

**Comment by alexander-beedie** [comment]
The `Roboto Mono` font for seems to have a slight kerning issue with our table reprs, leading to the edges all being slightly misaligned (increasing with the number of columns). Worth looking for a CSS fix (if possible) or maybe a different font/variation 🤔 

(Issue confirmed on two different OS/browser combinations).
